### ISSUE TITLE
Two question

### ISSUE THREAD

**Issue by huihuiyizhihua** [walidgeuttala/atsp_gnn#1]
The README file doesn’t seem to match the actual project/code
"How does this project ensure the generalizability of models trained on ATSP50 to ATSP datasets of different scales?"

**Comment by walidgeuttala** [comment]
Hi @huihuiyizhihua,

Thank you for your feedback. I have addressed your points below:

1. Project Structure and Workflows You were correct that the documentation needed improvement. I have updated README.md to accurately reflect the current project structure and valid workflows.

For specific details on data generation and running experiments, please refer to src/job_cpu.sh and src/job_gpu.sh. These scripts define the primary execution configurations and have been moved to the source directory for easier reference.

2. Model Generalization and Results Could you please clarify your second question? It is slightly vague, and I want to ensure I provide the specific details you need.

However, regarding the model performance:

Training: Our models are trained on instances of size 50. Being GNN-based, they can be tested on larger sizes.

Generalization: As you can see in the attached table image, we were capable of generalizing successfully to size 500.

Size 1000: We implemented a simplification technique for size 1000. It works well, but we have not yet beaten the GLOP baseline at that specific size, though we show strong performance at other sizes.

Our full paper is scheduled for publication in January, which will contain a complete breakdown of these results. In the meantime, I recommend reading the ATSP section of the [GLOP ](https://arxiv.org/abs/2312.08224) paper for more context on the baselines.

<img width="532" height="635" alt="Image" src="https://github.com/user-attachments/assets/98acc0a2-4806-4387-8425-6c8da0820021" />

**Comment by huihuiyizhihua** [comment]
Thanks for the prompt reply. I still have the following questions:
EB3O completeness: Has the full EB3O core algorithm been open-sourced? Only the skeleton is visible—are the key optimization strategies missing?
Training stability & generalization: Trained on 2 500 ATSP-50 and tested larger instances with batch_test_models; results vary wildly across runs and performance swings sharply during generalization. Could the issue be inconsistent dataset generation? A more detailed training config or any tricks to boost stability would be much appreciated.

**Comment by walidgeuttala** [comment]
EB3O is public in this repository:
https://github.com/akosholloszabo/regret_driven_heuristics

In the testing part of the code, you can see where the result files are saved and in which format. You need to provide the correct path to the saved files and update the file path in the repository accordingly. Specifically, please update the path in
EvaluateHeuristicsOnDifferentProblemSizesAndHyperParameters.kt, line 34.

In the coming weeks, I plan to integrate and transfer the current Kotlin code into Python so that the repository will become a complete Python implementation.

Regarding training stability and generalization: I believe there may be an issue in your current workflow. The dataset should be generated only once, then preprocessed and saved. After that, during training or testing, you should not regenerate the data, but instead reuse the saved dataset produced by:

src.data.dataset_generator

src.data.preprocessor

This should lead to consistent and stable results without unexpected variations. If this still does not work, I can run a concrete example and show you the exact steps.

Using the second repository after obtaining the test results should also be straightforward. If you observe inconsistencies, please describe exactly how you ran the script so I can understand what might be missing or misconfigured.

**Comment by huihuiyizhihua** [comment]
Thank you for your reply, author. I reran the process—could you please check if anything is missing or incorrectly configured?

Dataset download

<img width="1934" height="965" alt="Image" src="https://github.com/user-attachments/assets/016b7c44-3738-4968-be1d-d7f3a70f86f7" />

The training dataset is taken from the ATSP-50  of MatNet.
Sample 2 560 instances from the ATSP-50 set
   ls n50/*.atsp | shuf -n 2560 | xargs -I {} cp {} n50_selected/

Pre-processing
python src/data/dataset_generator.py \
  2560 50 saved_dataset/ATSP2560X50 \
  --regret_mode fixed_edge_lkh \
  --parallel \
  --lkh_path /LKH-3.0.9/LKH \
  --processes 42 \
  --from_tsplib_dir data/n50_selected

Split: 2 560 / 30 / 30
DGL_NO_GRAPHBOLT=1 python -m src.data.preprocessor \
  saved_dataset/ATSP_2560X50/ATSP_50x2560 \
  --atsp_size 50 \
  --relation_types ss st tt pp \
  --n_train 2500 --n_val 30 --n_test 30 \
  --save_pyg

Training
# Train HeteroGAT on ATSP-50
python -m src.engine.run \
  --mode train --framework dgl \
  --data_dir saved_dataset/ATSP_2560x50 \
  --model HetroGAT --atsp_size 50 \
  --batch_size 32 --n_epochs 100 \
  --agg concat --device cuda --undirected

<img width="2026" height="95" alt="Image" src="https://github.com/user-attachments/assets/38059ee2-a112-4572-99cd-286ca6e8375e" />

Hyper-parameter search
python -m src.engine.search_all_combo \
  --mode train --framework dgl \
  --data_dir saved_dataset/ATSP2560X50 \
  --model HetroGAT --atsp_size 50 \
  --batch_size 32 --n_epochs 100 \
  --lr_init 1e-3 --tb_dir ./runs \
  --device cuda --undirected \
  --n_trials 20 --seed 42 \
  --relation_types ss st tt pp \
  --relation_subsets "ss,st,tt,pp"

Testing (real ATSP-100 instances; sizes 150, 250, 500 generated on-the-fly)
python -m src.engine.batch_test_search_models

Test under training mode
atsp100
<img width="2026" height="36" alt="Image" src="https://github.com/user-attachments/assets/6a1b7512-6e05-421a-a195-c080b34f9242" />
atsp150
<img width="2030" height="31" alt="Image" src="https://github.com/user-attachments/assets/0837da79-fdf5-44f4-be8b-c41e28e5ce24" />
atsp250
<img width="2034" height="29" alt="Image" src="https://github.com/user-attachments/assets/75650687-3ba4-4ff3-aca3-5906eb6ab49f" />

atsp500

<img width="2045" height="49" alt="Image" src="https://github.com/user-attachments/assets/0193136f-4fdb-4d46-aa62-c5918aaa8651" />

Test under Hyper-parameter search mode
atsp100

<img width="2034" height="30" alt="Image" src="https://github.com/user-attachments/assets/8a052aed-b9ff-4944-bb15-b69a9ec061a5" />
atsp150

<img width="2059" height="35" alt="Image" src="https://github.com/user-attachments/assets/c5b0f11c-2783-425c-9de3-22c40c80b71d" />

atsp250

<img width="2034" height="30" alt="Image" src="https://github.com/user-attachments/assets/0776773e-65ca-41fa-95de-45b4be7afd0e" />

atsp500

<img width="2054" height="25" alt="Image" src="https://github.com/user-attachments/assets/1e608f07-1c32-47f4-a352-34f557f72799" />

Here is my full pipeline. I’m not sure which step is wrong, but the trained model generalizes poorly to larger ATSP instances. My doubts are:

1. During training, train/val/test are all done on ATSP-50. Could the optimizer be trapped in “small-graph local optima” that simply break when the graph gets bigger?

2. The metric used for training/validation is mis-aligned with the one used later for larger instances, so “better” numbers during training do not necessarily translate into better generalization.

3. When I re-sample a new random dataset , do I have to retune the hyper-parameters from scratch?

I look forward to your advice and guidance—thank you!

**Comment by huihuiyizhihua** [comment]
Hi @walidgeuttala, just a gentle follow-up on my last comment (2 days ago). Could you take a look when you have a moment? Happy to provide any extra logs or details if needed. Thanks!

**Comment by walidgeuttala** [comment]
Hi @huihuiyizhihua,

You did not use the code properly. For now, a quick solution to get results similar to my model is the following: if you solve the dataset the correct way for training, you should get identical results to those I showed in the table. Try this:

`python3 -m src.engine.run \
    --mode train \
    --framework dgl \
    --model HetroGAT \
    --data_dir ../saved_dataset/ATSP_3000x50 \
    --atsp_size 50 \
    --relation_types pp ss st tt \
    --agg concat \
    --hidden_dim 16 \
    --num_heads 1 \
    --num_gnn_layers 3 \
    --lr_init 0.0036302089692977634 \
    --batch_size 32 \
    --n_epochs 100 \
    --device cuda \
    --undirected \
    --seed 42



MODEL_PATH="Fill the model path"
# Test Size 100
python3 -m src.engine.run --mode test --framework dgl --atsp_size 100 --model HetroGAT --batch_size 1 \
    --data_path ../saved_dataset/ATSP_30x100 --tb_dir ./runs --agg concat \
    --model_path "$MODEL_PATH" \
    --device cuda --undirected --relation_types pp ss st tt
# Test Size 150
python3 -m src.engine.run --mode test --framework dgl --atsp_size 150 --model HetroGAT --batch_size 1 \
    --data_path ../saved_dataset/ATSP_30x150 --tb_dir ./runs --agg concat \
    --model_path "$MODEL_PATH" \
    --device cuda --undirected --relation_types pp ss st tt
# Test Size 250
python3 -m src.engine.run --mode test --framework dgl --atsp_size 250 --model HetroGAT --batch_size 1 \
    --data_path ../saved_dataset/ATSP_30x250 --tb_dir ./runs --agg concat \
    --model_path "$MODEL_PATH" \
    --device cuda --undirected --relation_types pp ss st tt
# Test Size 500
python3 -m src.engine.run --mode test --framework dgl --atsp_size 500 --model HetroGAT --batch_size 1 \
    --data_path ../saved_dataset/ATSP_30x500 --tb_dir ./runs --agg concat \
    --model_path "$MODEL_PATH" \
    --device cuda --undirected --relation_types pp ss st tt`


You may have some questions or need more details. I would be happy to help, but my work is not yet published. Once it is published, I will provide the full, proper code along with instructions for usage and achieving the best performance. That may take a few days, but as soon as everything is public, I will message you here for any follow-up or help you need.

**Comment by huihuiyizhihua** [comment]
Hi @walidgeuttala, , I reproduced the latest commands you provided, but when generalizing to larger scales, there is still a noticeable gap from the optimal solution (gap reaches 22.86% on atsp100, 22.56% on atsp150, and 39.22% on atsp250). I believe the performance difference is caused by discrepancies in the training data (after regenerating new training data, the original hyper-parameters may no longer be suitable). I suggest that you regenerate the training and test data to see whether the model remains effective, and that you open-source the exact data you used as well as the process that achieved the best results. Thank you, and I wish your paper a speedy acceptance.

test on atsp100
<img width="1671" height="30" alt="Image" src="https://github.com/user-attachments/assets/3490a21a-2876-4b27-9b70-097bc62d8062" />

test on atsp150
<img width="1695" height="30" alt="Image" src="https://github.com/user-attachments/assets/35776c07-5b12-4f52-90db-96eebe5a3ca9" />

test on atsp250
<img width="1675" height="30" alt="Image" src="https://github.com/user-attachments/assets/fac799a9-9935-48aa-952d-ecd31bf47112" />

**Comment by huihuiyizhihua** [comment]
Hi @walidgeuttala, "congratulations on the publication of your paper. Could you please make the complete training pipeline publicly available, as well as the data you generated? Thank you."

**Comment by huihuiyizhihua** [comment]
Hi @walidgeuttala, , I was wondering if you would be willing to disclose the training details or hyperparameter configurations?"

**Comment by walidgeuttala** [comment]
Hi,
​I would like to help, but between last week and this week, I’ve been
overloaded. I have high-priority stuff to finish first, but I have read
your previous email, and you are on my to-do list for this weekend

**Comment by walidgeuttala** [comment]
@huihuiyizhihua This is the link to the dataset I used: https://we.tl/t-GN7ZLlHXmt
I hope it works now. Please note that EB3O is not included here. You should obtain NN2R instead. To get EB3O, use the second repository I shared with you, which should provide the proper results. Please use the code correctly.

**Comment by walidgeuttala** [comment]
Hi @huihuiyizhihua, if it works for you please notify me so I close the issue.

**Comment by huihuiyizhihua** [comment]
Thank you for providing the data. I've been busy with other projects this week, and I'm planning to run through the data you gave me next week. I'll send you the results as soon as I get them. Thanks!

**Comment by huihuiyizhihua** [comment]
I noticed that the EB3O code repository shared earlier seems to be inaccessible. The author may have taken it down.

<img width="1879" height="405" alt="Image" src="https://github.com/user-attachments/assets/cdcb63a2-2b00-43e0-ac64-874431f99a45" />

**Comment by walidgeuttala** [comment]
Hi @huihuiyizhihua , this is the sharing link. It works for only three days on WeTransfer, so please download it as soon as possible
https://we.tl/t-vxXurUvF3k

**Comment by huihuiyizhihua** [comment]
<img width="1890" height="436" alt="Image" src="https://github.com/user-attachments/assets/c1768c14-9963-4478-af12-0d5100397345" />
Hi @huihuiyizhihua , I've downloaded the data connection. I wanted to mention that the previous eb30 code (https://github.com/akosholloszabo/regret_driven_heuristics) might have a broken link. Additionally, I ran the latest data and tested it , but the results still show significant discrepancies. I'll complete a full run-through as soon as possible and submit the complete training and testing process to you to see if there are any issues.

**Comment by walidgeuttala** [comment]
@huihuiyizhihua  I didn't quite understand your issue regarding the other repository. You should, of course, fix the paths to the predictions you generated. That repository hasn't been updated since I last used it, and it worked fine for me then, so it is likely that it isn't being used correctly.

On the other hand, it may contain a small bug, though I remember it wasn't difficult to run on my local machine. Since you mentioned that my code is producing different results, I will train one model, then test and validate it to see if I get the same results.

**Comment by walidgeuttala** [comment]
@huihuiyizhihua Did it work?

**Comment by huihuiyizhihua** [comment]
Hi @walidgeuttala,I apologize for the late reply"

**Comment by huihuiyizhihua** [comment]
git clone https://github.com/walidgeuttala/atsp_gnn.git

<img width="691" height="90" alt="Image" src="https://github.com/user-attachments/assets/0ff33a9d-cb28-4120-9b46-a0c6314d7cd3" />

python -m src.engine.run \
--mode train \
--framework dgl \
--model HetroGAT \
--data_dir /root/autodl-tmp/atsp_gnn/saved_dataset/ATSP_3000x50 \
--atsp_size 50 \
--relation_types pp ss st tt \
--agg concat \
--hidden_dim 16 \
--num_heads 1 \
--num_gnn_layers 3 \
--lr_init 0.0036302089692977634 \
--batch_size 32 \
--n_epochs 100 \
--device cuda \
--undirected \
--seed 42

<img width="691" height="166" alt="Image" src="https://github.com/user-attachments/assets/31c3be98-7b0e-4285-8f06-e18145addeba" />

ATSP500
python -m src.engine.run --mode test --framework dgl --atsp_size 500 --model HetroGAT --batch_size 1 \
--data_path /root/autodl-tmp/atsp_gnn/saved_dataset/ATSP_30x500 --tb_dir ./runs --agg concat \
--model_path "/root/autodl-tmp/atsp_gnn/runs/Mar03_16-26-51_HetroGATconcat_ATSP50_Combopp_ss_st_tt/trial_0/best_model.pt" \
--device cuda --undirected --relation_types pp ss st tt

<img width="691" height="74" alt="Image" src="https://github.com/user-attachments/assets/c38403fb-d69d-48e9-ab10-d0e9aded9b98" />


ATSP250
python -m src.engine.run --mode test --framework dgl --atsp_size 250 --model HetroGAT --batch_size 1 \
--data_path /root/autodl-tmp/atsp_gnn/saved_dataset/ATSP_30x250 --tb_dir ./runs --agg concat \
--model_path "/root/autodl-tmp/atsp_gnn/runs/Mar03_16-26-51_HetroGATconcat_ATSP50_Combopp_ss_st_tt/trial_0/best_model.pt" \
--device cuda --undirected --relation_types pp ss st tt

<img width="691" height="73" alt="Image" src="https://github.com/user-attachments/assets/b53952cf-204c-4745-acc0-67325b2c97f0" />

ATSP150

<img width="692" height="74" alt="Image" src="https://github.com/user-attachments/assets/b3affe64-b130-4881-9001-d399f0460861" />



ATSP50

<img width="690" height="74" alt="Image" src="https://github.com/user-attachments/assets/debf9ebc-8a6d-498b-9d84-62ec7d190fbc" />

The GAP results obtained using the parameters you provided in your comment
ATSP500:18.71
ATSP250:13.99
ATSP150:11.57
ATSP100:10.68
The results are close to those in the paper


However, the model trained with the parameters reported in the paper (batch size 15, learning rate 0.001)"

python -m src.engine.run \
--mode train \
--framework dgl \
--model HetroGAT \
--atsp_size 50 \
--relation_types pp ss st tt \
--agg concat \
--lr_init 0.001 \
--batch_size 15 \
--undirected \
--device cuda \
--data_dir /root/autodl-tmp/atsp_gnn/saved_dataset/ATSP_3000x50 

<img width="690" height="72" alt="Image" src="https://github.com/user-attachments/assets/cb0c0022-8a57-477f-a127-f9783ac2d031" />

<img width="692" height="69" alt="Image" src="https://github.com/user-attachments/assets/ef2639cf-a9af-4dc7-a903-4d4cf7594faa" />

Atsp500:43.49
Atsp250:20.50
My current question is: whenever there is any discrepancy between the parameters and those you provided in your comment, the performance drops significantly when generalizing to large-scale testing

Is it only under these specific parameters that a model with good performance can be trained?


"Additionally, the currently shared code only includes the NN2R component, and the EB3O code is not present. The statement 'EB3O is public in this repository: https://github.com/akosholloszabo/regret_driven_heuristics' could not be verified, as the link is inaccessible."

**Comment by walidgeuttala** [comment]
Hi @huihuiyizhihua I couldn't view the photos you mentioned. It's possible there is a discrepancy in our setups leading to the varied results, and running a thorough hyperparameter search might help you find better configurations. On another note, I spoke with the repository owner, and he made it public I hope it is all fine now.
